src/Shared/ServerInfrastructure/HttpCharacters.cs

```diff
@@ -1,224 +1,54 @@
+    // user:password@host.com
+    private static readonly IndexOfAnyValues<byte> _allowedAuthorityBytes = IndexOfAnyValues.Create(":.-[]@0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"u8);
 
-    private static bool[] InitializeAlphaNumeric()
-    {
-        // ALPHA and DIGIT https://tools.ietf.org/html/rfc5234#appendix-B.1
-        var alphaNumeric = new bool[_tableSize];
-        for (var c = '0'; c <= '9'; c++)
-        {
-            alphaNumeric[c] = true;
-        }
-        for (var c = 'A'; c <= 'Z'; c++)
-        {
-            alphaNumeric[c] = true;
-        }
-        for (var c = 'a'; c <= 'z'; c++)
-        {
-            alphaNumeric[c] = true;
-        }
-        return alphaNumeric;
-    }
+    // Matches Http.Sys
+    // Matches RFC 3986 except "*" / "+" / "," / ";" / "=" and "%" HEXDIG HEXDIG which are not allowed by Http.Sys
+    private static readonly IndexOfAnyValues<char> _allowedHostChars = IndexOfAnyValues.Create("!$&'()-._~" + AlphaNumeric);
 
-    private static bool[] InitializeAuthority()
-    {
-        // Authority https://tools.ietf.org/html/rfc3986#section-3.2
-        // Examples:
-        // microsoft.com
-        // hostname:8080
-        // [::]:8080
-        // [fe80::]
-        // 127.0.0.1
-        // user@host.com
-        // user:password@host.com
-        var authority = new bool[_tableSize];
-        Array.Copy(_alphaNumeric, authority, _tableSize);
-        authority[':'] = true;
-        authority['.'] = true;
-        authority['-'] = true;
-        authority['['] = true;
-        authority[']'] = true;
-        authority['@'] = true;
-        return authority;
-    }
+    // tchar https://tools.ietf.org/html/rfc7230#appendix-B
+    private static readonly IndexOfAnyValues<char> _allowedTokenChars = IndexOfAnyValues.Create("!#$%&'*+-.^_`|~" + AlphaNumeric);
+    private static readonly IndexOfAnyValues<byte> _allowedTokenBytes = IndexOfAnyValues.Create("!#$%&'*+-.^_`|~0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"u8);
 
-    private static bool[] InitializeToken()
-    {
-        // tchar https://tools.ietf.org/html/rfc7230#appendix-B
-        var token = new bool[_tableSize];
-        Array.Copy(_alphaNumeric, token, _tableSize);
-        token['!'] = true;
-        token['#'] = true;
-        token['$'] = true;
-        token['%'] = true;
-        token['&'] = true;
-        token['\''] = true;
-        token['*'] = true;
-        token['+'] = true;
-        token['-'] = true;
-        token['.'] = true;
-        token['^'] = true;
-        token['_'] = true;
-        token['`'] = true;
-        token['|'] = true;
-        token['~'] = true;
-        return token;
-    }
+    // field-value https://tools.ietf.org/html/rfc7230#section-3.2
+    // HTAB, [VCHAR, SP]
+    private static readonly IndexOfAnyValues<char> _allowedFieldChars = IndexOfAnyValues.Create("\t !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~");
```

Comment:
Nit: since you're already using the constant elsewhere
```suggestion
    private static readonly IndexOfAnyValues<char> _allowedFieldChars = IndexOfAnyValues.Create("\t !\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~" + AlphaNumeric);
```

This file is quite beautiful now 🎉

Reply:
I knew that I'll love this API from the beginning, but it really shines when I see how clean the code becomes.

Reply:
C# needs to enable something like `$"\t\n{AlphaNumeric}"u8` then it's even cleaner for `char` and `byte` as it can be re-used.
Or that for the utf8-literals a constant string can be (mis-) used. I think that could work, as at parsing time the constant is present.

Are you aware of some issues for this? Or is that too niche?

Reply:
Not aware of an existing issue for it, but I agree it would come in handy.
Another example is where you have the same constant for chars and bytes - e.g. in
https://github.com/dotnet/runtime/blob/77de7840c7ad806cd5ad5dc9555ad3d3bed2a6e5/src/libraries/System.Net.Http/src/System/Net/Http/HttpRuleParser.cs#L14-L18 as well